galaxy_importer/ansible_test/builders/pulp_build.py

```diff
@@ -15,7 +15,271 @@
 # You should have received a copy of the Apache License
 # along with Galaxy.  If not, see <http://www.apache.org/licenses/>.
 
+import json
+import logging
+import os
+import requests
+import subprocess
+import time
+import uuid
 
-def build_image_with_artifact():
+from galaxy_importer import config
+from galaxy_importer import constants
+from galaxy_importer import exceptions
+from shutil import copy, copyfile, copyfileobj
+
+default_logger = logging.getLogger(__name__)
+config_data = config.ConfigFile.load()
+cfg = config.Config(config_data=config_data)
+API_CHECK_RETRIES = 360
+API_CHECK_DELAY_SECONDS = 1
+
+
+class Build(object):
     """Use pulp-container to build ansible-test image with artifact inside."""
-    return ''
+    def __init__(self, pulp_artifact_file, api_url, content_url, metadata, logger):
+        self.api_url = api_url
+        self.content_url = content_url
+        self.basic_auth = ('admin', 'admin')
+        self.pulp_artifact_file = pulp_artifact_file
+        self.metadata = metadata
+        self.shared_dockerfile = os.path.join(
+            os.path.dirname(constants.ROOT_DIR.rstrip('galaxy_importer')),
+            'docker/ansible-test/Dockerfile')
+        self.dockerfile = os.path.join(
+            os.path.dirname(os.path.realpath(__file__)),
+            'Dockerfile')
+        self.collection_file = self._get_collection_file()
+        self.registry_base_path = 'galaxy-importer'
+        self.registry_name = 'archive-test'
+        self.log = logger or default_logger
+        self.setup_dockerfile()
+
+    def _get_collection_file(self):
+        path = os.path.join(
+            os.path.dirname(os.path.realpath(__file__)),
+            'archive.tar.bz')
+        try:
+            url = Build._get_pulp_archive_url(self.pulp_artifact_file)
+            with requests.get(url, stream=True) as r:
+                with open(path, 'wb') as f:
+                    copyfileobj(r.raw, f)
+        except AttributeError:
```

Comment:
there is a lot of logic in this exception, is this an expected code path? when would the exception occur / not occur?